### ISSUE TITLE
Go: displayed types are not complete

### ISSUE THREAD

**Issue by disconnect3d** [semgrep/semgrep#1768]
Not sure if that is a bug, but the displayed types by `message` key in the rule format are a bit incomplete.

PoC:
```
$ cat a.yml
rules:
- id: some-rule
  pattern: func ($O *$TYPE) $METHOD($A *$TA, $B *$TB) { ... }
  message: arg1=`$TA` arg2=`$TB`
  severity: WARNING
  languages: [go]

$ cat a.go
package heh
func (o *T) Foo(a *some.NamespacedType, b *some.OtherType) { }
func (o *T) Bar(a *interface{}, b *struct{}) { return }

$ semgrep -f a.yml | grep severity
running 1 rules...
severity:warning rule:some-rule: arg1=`some NamespacedType` arg2=`some OtherType`
severity:warning rule:some-rule: arg1=`interface` arg2=`{}`
```

There are three weird cases:
1. The `some.NamespacedType` is displayed without dots, as `some NamespacedType`
2. The `interface{}` is displayed as just `interface`
3. The `struct{}` is displayed as `{}`

**Comment by brendongo** [comment]
Convenience repro link: https://semgrep.dev/s/QwNJ

Checking metavars locally with --json has output: 
```
"metavars": {
          "$TB": {
            "start": {
              "line": 2,
              "col": 44,
              "offset": 55
            },
            "end": {
              "line": 2,
              "col": 58,
              "offset": 69
            },
            "abstract_content": "some OtherType",
            "unique_id": {
              "type": "AST",
              "md5sum": "93c4313b9bdeb983dad54c719c7c7dff"
            }
          },
          "$B": {
            "start": {
              "line": 2,
              "col": 41,
              "offset": 52
            },
            "end": {
              "line": 2,
              "col": 42,
              "offset": 53
            },
            "abstract_content": "b",
            "unique_id": {
              "type": "id",
              "value": "b",
              "kind": "Param",
              "sid": 2
            }
          },
          "$TA": {
            "start": {
              "line": 2,
              "col": 20,
              "offset": 31
            },
            "end": {
              "line": 2,
              "col": 39,
              "offset": 50
            },
            "abstract_content": "some NamespacedType",
            "unique_id": {
              "type": "AST",
              "md5sum": "2af73acda8772054a355b37e8fa59db3"
            }
          },
          "$A": {
            "start": {
              "line": 2,
              "col": 17,
              "offset": 28
            },
            "end": {
              "line": 2,
              "col": 18,
              "offset": 29
            },
            "abstract_content": "a",
            "unique_id": {
              "type": "id",
              "value": "a",
              "kind": "Param",
              "sid": 1
            }
          },
```

Looks like the . and brackets are being stripped by semgrep-core matcher.

Full --json output for ease in debugging:
```
{
  "results": [
    {
      "check_id": "some-rule",
      "path": "b.go",
      "start": {
        "line": 2,
        "col": 7
      },
      "end": {
        "line": 2,
        "col": 63
      },
      "extra": {
        "message": "arg1=`some NamespacedType` arg2=`some OtherType`\n",
        "metavars": {
          "$TB": {
            "start": {
              "line": 2,
              "col": 44,
              "offset": 55
            },
            "end": {
              "line": 2,
              "col": 58,
              "offset": 69
            },
            "abstract_content": "some OtherType",
            "unique_id": {
              "type": "AST",
              "md5sum": "93c4313b9bdeb983dad54c719c7c7dff"
            }
          },
          "$B": {
            "start": {
              "line": 2,
              "col": 41,
              "offset": 52
            },
            "end": {
              "line": 2,
              "col": 42,
              "offset": 53
            },
            "abstract_content": "b",
            "unique_id": {
              "type": "id",
              "value": "b",
              "kind": "Param",
              "sid": 2
            }
          },
          "$TA": {
            "start": {
              "line": 2,
              "col": 20,
              "offset": 31
            },
            "end": {
              "line": 2,
              "col": 39,
              "offset": 50
            },
            "abstract_content": "some NamespacedType",
            "unique_id": {
              "type": "AST",
              "md5sum": "2af73acda8772054a355b37e8fa59db3"
            }
          },
          "$A": {
            "start": {
              "line": 2,
              "col": 17,
              "offset": 28
            },
            "end": {
              "line": 2,
              "col": 18,
              "offset": 29
            },
            "abstract_content": "a",
            "unique_id": {
              "type": "id",
              "value": "a",
              "kind": "Param",
              "sid": 1
            }
          },
          "$TYPE": {
            "start": {
              "line": 2,
              "col": 10,
              "offset": 21
            },
            "end": {
              "line": 2,
              "col": 11,
              "offset": 22
            },
            "abstract_content": "T",
            "unique_id": {
              "type": "AST",
              "md5sum": "49c740832a089b004a2456837c76b63e"
            }
          },
          "$O": {
            "start": {
              "line": 2,
              "col": 7,
              "offset": 18
            },
            "end": {
              "line": 2,
              "col": 8,
              "offset": 19
            },
            "abstract_content": "o",
            "unique_id": {
              "type": "AST",
              "md5sum": "63d677fd2c443ef3f4688b37e5673b89"
            }
          },
          "$METHOD": {
            "start": {
              "line": 2,
              "col": 13,
              "offset": 24
            },
            "end": {
              "line": 2,
              "col": 16,
              "offset": 27
            },
            "abstract_content": "Foo",
            "unique_id": {
              "type": "AST",
              "md5sum": "1f0c9b0d0f3f351a3dd7a4e7ca77c675"
            }
          }
        },
        "metadata": {},
        "severity": "WARNING",
        "lines": "func (o *T) Foo(a *some.NamespacedType, b *some.OtherType) { }"
      }
    },
    {
      "check_id": "some-rule",
      "path": "b.go",
      "start": {
        "line": 3,
        "col": 7
      },
      "end": {
        "line": 3,
        "col": 56
      },
      "extra": {
        "message": "arg1=`interface` arg2=`{}`\n",
        "metavars": {
          "$TB": {
            "start": {
              "line": 3,
              "col": 42,
              "offset": 116
            },
            "end": {
              "line": 3,
              "col": 44,
              "offset": 118
            },
            "abstract_content": "{}",
            "unique_id": {
              "type": "AST",
              "md5sum": "d932b62a9c2ba8dc018d4d910f42d2d4"
            }
          },
          "$B": {
            "start": {
              "line": 3,
              "col": 33,
              "offset": 107
            },
            "end": {
              "line": 3,
              "col": 34,
              "offset": 108
            },
            "abstract_content": "b",
            "unique_id": {
              "type": "id",
              "value": "b",
              "kind": "Param",
              "sid": 4
            }
          },
          "$TA": {
            "start": {
              "line": 3,
              "col": 20,
              "offset": 94
            },
            "end": {
              "line": 3,
              "col": 29,
              "offset": 103
            },
            "abstract_content": "interface",
            "unique_id": {
              "type": "AST",
              "md5sum": "10d2b392eff673aa9885191cdf868951"
            }
          },
          "$A": {
            "start": {
              "line": 3,
              "col": 17,
              "offset": 91
            },
            "end": {
              "line": 3,
              "col": 18,
              "offset": 92
            },
            "abstract_content": "a",
            "unique_id": {
              "type": "id",
              "value": "a",
              "kind": "Param",
              "sid": 3
            }
          },
          "$TYPE": {
            "start": {
              "line": 3,
              "col": 10,
              "offset": 84
            },
            "end": {
              "line": 3,
              "col": 11,
              "offset": 85
            },
            "abstract_content": "T",
            "unique_id": {
              "type": "AST",
              "md5sum": "49c740832a089b004a2456837c76b63e"
            }
          },
          "$O": {
            "start": {
              "line": 3,
              "col": 7,
              "offset": 81
            },
            "end": {
              "line": 3,
              "col": 8,
              "offset": 82
            },
            "abstract_content": "o",
            "unique_id": {
              "type": "AST",
              "md5sum": "63d677fd2c443ef3f4688b37e5673b89"
            }
          },
          "$METHOD": {
            "start": {
              "line": 3,
              "col": 13,
              "offset": 87
            },
            "end": {
              "line": 3,
              "col": 16,
              "offset": 90
            },
            "abstract_content": "Bar",
            "unique_id": {
              "type": "AST",
              "md5sum": "523dbe3a8ea8b69f369b82bd1dbe6d45"
            }
          }
        },
        "metadata": {},
        "severity": "WARNING",
        "lines": "func (o *T) Bar(a *interface{}, b *struct{}) { return }"
      }
    }
  ],
  "errors": []
}
```

**Comment by disconnect3d** [comment]
Ups, found another case that is worse than previous (since we can infer `struct{}` from `{}` and `interface{}` from `interface` but here we can't):

```
$ cat a.yml
rules:
- id: some-rule
  pattern: func ($O *$TYPE) $METHOD($A *$TA, $B *$TB) { ... }
  message: arg1=`$TA` arg2=`$TB`
  severity: WARNING
  languages: [go]

$ cat a.go
package heh
func (o *T) Baz(a *interface{}, b *[]string) { return }

$ semgrep -f a.yml | grep severity
running 1 rules...
severity:warning rule:some-rule: arg1=`interface` arg2=`string`
```

TL;DR: The `[]string` was displayed as `string`.

**Comment by aryx** [comment]
Maybe it's time for the python wrapper to not use the abstract_content field, which as the name suggest results from the Abstract Syntax Tree which does not have the full syntactic information that you would find in a concrete syntax tree.

**Comment by aryx** [comment]
I can improve a bit the case for []string and struct vs interface but not much more.

**Comment by aryx** [comment]
Or we should use @emjin pretty printer when printing the message containing metavariable.

**Comment by brendongo** [comment]
Oh I see. Yeah makes sense to use range instead of abstract_content field when figuring out metavariable values/autofixes.

**Comment by brendongo** [comment]
This issue re inclusive/exclusive is probably relevant if we switch over to using ranges: https://github.com/returntocorp/semgrep/issues/1721#issuecomment-702338846

**Comment by aryx** [comment]
Works better now.
https://semgrep.dev/s/7g2G/?version=develop
still not perfect, but at least we have now struct{}, and string[].
Is it enough for now @disconnect3d ? The right fix requires to switch to a different way to print metavariables in message, but that make take a while to fix.

**Comment by aryx** [comment]
I'll close for now, we can create another issue to really use pretty printing when displaying metavar.

**Comment by brendongo** [comment]
FYI @aryx when switching to using line ranges in #1859 we now correctly handle case 1 and 3 of the user above but case 2 (interface{}) is still reported only as (interface) seems like this is semgrep-core not including the {} in the case of interface
